CRM/Contribute/Form/Task/Invoice.php

```diff
@@ -543,7 +543,12 @@ public static function printPDF($contribIDs, &$params, $contactIds) {
         self::addActivities($subject, $contribution->contact_id, $fileName, $params);
       }
 
-      CRM_Core_DAO::setFieldValue('CRM_Contribute_DAO_Contribution', $contribution->id, 'invoice_id', $invoiceId);
+      // CRM-20387 save my human-readable invoice number as invoice_id in the contribution table,
+      // but only if the current value is one of: set by this invoicing code, or pay-later, or empty.
+      $set_by_invoicing_code = (0 === strpos($contribution->invoice_id, CRM_Utils_Array::value('invoice_prefix', $prefixValue)));
```

Comment:
Invoice prefixes are designed to be configurable, so this is not a reliable algorithm. Were someone to make use of that functionality, they would find their invoice numbers being overwritten. On a bug fix and coding approaching I'm not keen on breaking other stuff to fix a problem, so I still think a proper permanent fix is the better choice here. However, I don't have a dog in this hunt though - I'll let CompuCorp decide what they think.